Dynamic Branding Name Support Added For Export File Name Prefix

shamim-emon opened the pull request:
Please ensure that your pull request meets the following requirements - thanks!

- Does not contain merge commits. Rebase instead.
- Contains commits with descriptive titles.
- New code is written in Kotlin whenever possible.
- Follows our existing codestyle (`gradlew spotlessCheck` to check and `gradlew spotlessFormat` to format your source code; will be checked by CI).
- Does not break any unit tests (`gradlew testDebugUnitTest`; will be checked by CI).
- Uses a descriptive title; don't put issue numbers in there.
- Contains a reference to the issue that it fixes (e.g. _Closes #XXX_ or _Fixes #XXX_) in the body text.
- For cosmetic changes add one or multiple images, if possible.

Finally, please replace this template text with a description of the change and additional context if necessary.

cketti commented:
I don't think the brand name is a good fit for this. "K-9 Mail" contains a space and some translations even localize it.

The prefix for K-9 Mail should remain "k9". The prefix for Thunderbird can be "thunderbird".

I suggest adding a `FilePrefixProvider` interface. For now it can live in the package `com.fsck.k9.preferences`.

`K9AppNameProvider` and `TbAppNameProvider` could then also implement this interface and provide the language-independent values mentioned above.

shamim-emon commented:
> I don't think the brand name is a good fit for this. "K-9 Mail" contains a space and some translations even localize it.
> 
> The prefix for K-9 Mail should remain "k9". The prefix for Thunderbird can be "thunderbird".
> 
> I suggest adding a `FilePrefixProvider` interface. For now it can live in the package `com.fsck.k9.preferences`.
> 
> `K9AppNameProvider` and `TbAppNameProvider` could then also implement this interface and provide the language-independent values mentioned above.

@cketti done! Please let me know if anything else needed. Thanks

cketti commented on app-k9mail/src/main/kotlin/app/k9mail/provider/K9AppNameProvider.kt:
```diff
@@ -4,15 +4,26 @@ import android.content.Context
+
+    override val filePrefix: String
+        get() {
+            val now = Calendar.getInstance()
+            val dateFormat = SimpleDateFormat("yyyy-MM-dd", Locale.US)
```
This logic should stay in `SettingsExporter`. The prefix supplied by `FilePrefixProvider` should only be "k9" and "thunderbird".

shamim-emon replied:
@cketti Done!

cketti commented on app-k9mail/src/main/kotlin/app/k9mail/provider/K9AppNameProvider.kt:
```diff
@@ -4,15 +4,20 @@ import android.content.Context
         context.getString(R.string.app_name)
     }
+
+    override val filePrefix: String by lazy {
+        "k9"
+    }
```
There's nothing that would benefit from computing this value lazily. So this can simply be:

```suggestion
    override val filePrefix: String = "k9"
```

cketti commented on app-thunderbird/src/main/kotlin/net/thunderbird/android/provider/TbAppNameProvider.kt:
```diff
@@ -3,16 +3,21 @@ package net.thunderbird.android.provider
         context.getString(R.string.brand_name)
     }
+
+    override val filePrefix: String by lazy {
+        "thunderbird"
+    }
```
```suggestion
    override val filePrefix: String = "thunderbird"
```

cketti commented on legacy/core/src/main/java/com/fsck/k9/preferences/SettingsExporter.kt:
```diff
@@ -482,11 +483,10 @@ class SettingsExporter(
         val now = Calendar.getInstance()
         val dateFormat = SimpleDateFormat("yyyy-MM-dd", Locale.US)
-        return String.format("%s_%s.%s", EXPORT_FILENAME_PREFIX, dateFormat.format(now.time), EXPORT_FILENAME_SUFFIX)
+        return String.format("%s_%s.%s", filePrefixProvider.filePrefix, dateFormat.format(now.time), EXPORT_FILENAME_SUFFIX)
```
Now, instead of `thunderbird_settings_export_2024-10-22.k9s`, we'll end up with `thunderbird_2024-10-22.k9s`.

```suggestion
        return String.format("%s_%s_%s.%s", EXPORT_FILENAME_MIDDLE_PART, filePrefixProvider.filePrefix, dateFormat.format(now.time), EXPORT_FILENAME_SUFFIX)
```

With
```kotlin
companion object {
    private const val EXPORT_FILENAME_MIDDLE_PART = "settings_export"
    // …
}
```

shamim-emon replied:
@cketti there is an issue with argument order. Fixed it now.

cketti commented:
Looks like there are only code style issues left. You can fix these by running `./gradlew spotlessApply`

shamim-emon commented:
> Looks like there are only code style issues left. You can fix these by running `./gradlew spotlessApply`

@cketti spotless issues should be fixed now.

shamim-emon commented:
> > Looks like there are only code style issues left. You can fix these by running `./gradlew spotlessApply`
> 
> @cketti spotless issues are fixed now. Fixed the style issues with `Spotless Gradle` plugin(As some configuration needs to be done in my device before I could run above command from terminal and get result). Please allow the workflow execution. In-case of failure I will do the necessary configuration and run `./gradlew spotlessApply`

@cketti there was an detekt linting issue that I have fixed. for future reference could you please point me to the docs/ provide list to checks I can perform locally to ensure all checks pass in CI?
Thanks

cketti commented:
Thanks :+1:

You can see what checks are run here: https://github.com/thunderbird/thunderbird-android/blob/56eaf0a559cd55b018fdd9987a4b29431b5d2887/.github/workflows/android.yml#L40-L56
